processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java

```diff
@@ -0,0 +1,106 @@
+
+public class Issue1801BuilderProvider extends ImmutablesBuilderProvider implements BuilderProvider {
+
+    private BuilderInfo findBuilderInfoForImmutables(TypeElement typeElement) {
+        return findBuilderInfoFromInnerBuilderClass( typeElement );
+    }
+
+    /**
+     * Looks for inner builder class in the Immutable interface / abstract class.
+     *
+     * The inner builder class should be be declared with the following line
+     *
+     * <pre>
+     *     public static Builder() extends ImmutableItem.Builder { }
+     * </pre>
+     *
+     * The Immutable instance should be created with the following line
+     *
+     * <pre>
+     *     new Item.Builder().withId("123").build();
+     * </pre>
+     *
+     * @see org.mapstruct.ap.test.bugs._1801.domain.Item
+     *
+     * @param typeElement
+     * @return
+     */
+    private BuilderInfo findBuilderInfoFromInnerBuilderClass(TypeElement typeElement) {
+        if (shouldIgnore( typeElement )) {
+            return null;
+        }
+
+        List<TypeElement> types = ElementFilter.typesIn( typeElement.getEnclosedElements() );
+        List<BuilderInfo> builderInfo = new ArrayList<>();
+        types.stream()
+             .filter( innerType -> isPossibleInnerBuilder( innerType, typeElement ) )
+             .forEach( innerBuilderClass -> {
+                 Collection<ExecutableElement> buildMethods = findBuildMethods( innerBuilderClass, typeElement );
+                 if (!buildMethods.isEmpty()) {
+                     ElementFilter.constructorsIn( innerBuilderClass.getEnclosedElements() )
+                                  .stream()
+                                  .filter( constructor -> constructor.getParameters().isEmpty() &&
+                                      constructor.getModifiers().contains( Modifier.PUBLIC ) )
+                                  .findAny()
+                                  .ifPresent( defaultConstructor -> builderInfo.add(
+                                      new BuilderInfo.Builder()
+                                          .builderCreationMethod( defaultConstructor )
+                                          .buildMethod( buildMethods )
+                                          .build()
+                                  ) );
+                 }
+             } );
+
+        if (builderInfo.size() == 1) {
+            return builderInfo.get( 0 );
+        }
+        else if (builderInfo.size() > 1) {
+            throw new MoreThanOneBuilderCreationMethodException(typeElement.asType(), builderInfo);
+        }
+
+        return null;
+    }
+
+    private boolean isPossibleInnerBuilder(TypeElement innerClass, TypeElement outerClass) {
+        final String packageName = this.elementUtils.getPackageOf( outerClass ).getQualifiedName().toString();
+        final Name outerSimpleName = outerClass.getSimpleName();
+        final String builderClassName = packageName + ".Immutable" + outerSimpleName + ".Builder";
```

Comment:
Do we really need modificator  "final" for variables: packageName, outerSimpleName, builderClassName?